### ISSUE TITLE
EKS authentication slower with aws-cli than aws-iam-authenticator

### ISSUE THREAD

**Issue by emaincourt** [aws/aws-cli#5570]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [X] I've gone though the [User Guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-welcome.html) and the [API reference](https://docs.aws.amazon.com/cli/latest/reference/)
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-cli/issues) and didn't find any solution
  
**Describe the bug**
When using the `aws-cli`, it takes around 5 seconds to get a proper token to communicate with our EKS clusters. Performing the same action with `aws-iam-authenticator` takes less than a second.

**SDK version number**
`aws-cli/2.0.50`
`aws-iam-authenticator/0.5.1`

**Platform/OS/Hardware/Device**
Both MacOS and Ubuntu

**To Reproduce (observed behavior)**
```
time aws-iam-authenticator token -i eks-staging                                                                                                                                                                                         (arn:aws:eks:eu-west-3:<REDACTED>:cluster/eks-staging/default)
{"kind":"ExecCredential","apiVersion":"client.authentication.k8s.io/v1alpha1","spec":{},"status":{"expirationTimestamp":"2020-09-24T13:53:42Z","token":"<REDACTED>"}}
aws-iam-authenticator token -i eks-staging  0.09s user 0.05s system 14% cpu 0.966 total

time aws eks get-token --cluster-name eks-staging                                                                                                                                                                                       (arn:aws:eks:eu-west-3:<REDACTED>:cluster/eks-staging/default)
{"kind": "ExecCredential", "apiVersion": "client.authentication.k8s.io/v1alpha1", "spec": {}, "status": {"expirationTimestamp": "2020-09-24T14:03:47Z", "token": "<REDACTED>"}}
aws eks get-token --cluster-name eks-staging  0.71s user 0.23s system 21% cpu 4.347 total
```

**Expected behavior**
Both should respond in the same time.

**Logs/output**

**Additional context**

**Comment by kdaily** [comment]
Hi @emaincourt, thanks for reporting.

This sounds like an enhancement, not a bug. We would have to investigate what the implementation of `aws-iam-authenticator` does.

Have you tested this in AWS CLI v1 to see if this is a change introduced in the AWS CLI v2?

**Comment by nielsole** [comment]
```
$ aws eks get-token --cluster-name [...]
real	0m2,565s
user	0m0,515s
sys	0m0,049s

$ time aws-iam-authenticator token -i [...]
real	0m0,019s
user	0m0,014s
sys	0m0,007s

$ aws --version
aws-cli/2.0.59 Python/3.7.3 Linux/5.8.16-050816-generic exe/x86_64.ubuntu.20
```
Just form the timing it looks like aws CLI is fetching a new token, while aws-iam-authenticator is just reusing my existing one.

**Comment by nielsole** [comment]
Looks like it tries to contact metadata server twice with a 1 second timeout.
```
[pid 23847] connect(5, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("169.254.169.254")}, 16) = -1 EINPROGRESS (Operation now in progress)
[pid 23847] clock_gettime(CLOCK_MONOTONIC, {tv_sec=13203, tv_nsec=435005554}) = 0
[pid 23847] poll([{fd=5, events=POLLOUT|POLLERR}], 1, 1000

) = 0 (Timeout)
[pid 23847] close(5)                    = 0
[pid 23847] clock_gettime(CLOCK_MONOTONIC, {tv_sec=13204, tv_nsec=437736837}) = 0
[pid 23847] socket(AF_INET, SOCK_STREAM, IPPROTO_TCP) = 5
[pid 23847] ioctl(5, FIOCLEX)           = 0
[pid 23847] ioctl(5, FIONBIO, [1])      = 0
[pid 23847] connect(5, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("169.254.169.254")}, 16) = -1 EINPROGRESS (Operation now in progress)
[pid 23847] clock_gettime(CLOCK_MONOTONIC, {tv_sec=13204, tv_nsec=438882515}) = 0
[pid 23847] poll([{fd=5, events=POLLOUT|POLLERR}], 1, 1000

) = 0 (Timeout)
[pid 23847] close(5)   
```

**Comment by nielsole** [comment]
Setting the `AWS_EC2_METADATA_DISABLED=true` (undocumented?) from https://github.com/aws/aws-cli/issues/5623 reduces query aws eks get-token to ~500ms (which is still slow, but not bug-level slow ;) )

**Comment by stephank** [comment]
The metadata server stuff is interesting. The other major factor is just loading all of the Python code, I think. I time ~0.85s for `aws eks get-token ...`, compared to ~0.75s for just `aws` (which only prints usage). A bare `print('hello')` in my Python is ~0.05s. Solving this sounds tricky, but I don't know Python well, maybe there are some magic ways to tune this?

Maybe another solution is to add a flag to `aws eks update-kubeconfig` to have it generate a config that uses aws-iam-authenticator? Or perhaps even default to it, if it is found in `$PATH`?

**Comment by chris-oconnor-plurilock** [comment]
I also have the exact same problem. We have just moved over to aws sso so we are forced to use aws eks get-token if we use aws sso to gain cli access to the eks cluster. So I would suggest this is actually a regression since i can no longer use aws-iam-authenticator. The speed difference is very annoying

**Comment by dhensen** [comment]
I ended up here after investigating why my pytest took so long to initialize. First I thought it was WSL on a corporate laptop, but it was as slow on my own 5950X. It turns out calls to that IP address mentioned in comments above add 8 to 10 seconds to my pytest initialization. I'm using boto3. boto3 does stuff with aws-cli. So I'm still not 100% sure, but at least adding AWS_EC2_METADATA_DISABLED=true gives some relieve. I do wonder what I am now missing out on...?

**Comment by aogier** [comment]
I'm facing the same problem and I see that, for reasons beyond my comprehension, `apt-cache policy` gets called at every aws invocation, including get-token:

```
strace -fe file aws ecr describe-repositories 2>&1 | grep apt-cache
[pid 213326] execve("/home/oggei/.pyenv/shims/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/home/oggei/.pyenv/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/home/oggei/.zinit/polaris/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/home/oggei/.poetry/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/home/oggei/.local/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/home/oggei/.poetry/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/home/oggei/.cargo/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/home/oggei/.local/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/usr/local/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = -1 ENOENT (No such file or directory)
[pid 213326] execve("/usr/bin/apt-cache", ["apt-cache", "policy"], 0x7f1c6ee78210 /* 58 vars */) = 0
```
apt-cache policy is indeed slow:
```
time apt-cache policy > /dev/null
apt-cache policy > /dev/null  5.18s user 0.23s system 100% cpu 5.409 total
```

and so is any kubectl command:
```
time k get no > /dev/null
apt-cache policy > /dev/null  5.56s user 0.23s system 100% cpu 5.710 total
```

**Comment by sov-msaeed** [comment]
I didn't have this problem but recently, it takes almost 4 second to get the token. I don't know what has changed, was there any change recently ?

```
aws --version
aws-cli/2.30.2 Python/3.13.7 Darwin/24.6.0 source/arm64
```

**Comment by PadenZach** [comment]
These seems to be a small part of the bottleneck for me.

I wrote a script to replace the existing get-token method with one that would only call it if the cached token had expired (just a bit of bash and jq).

I got it to successfully work, but I dont think the speed of get token is important here.

I tested `kubectl get svc` against a demo EKS cluster with just a handful of services.

Without caching:
```
 hyperfine 'kubectl get svc'
Benchmark 1: kubectl get svc
  Time (mean ± σ):      6.073 s ±  0.086 s    [User: 0.296 s, System: 0.172 s]
  Range (min … max):    5.935 s …  6.169 s    10 runs
 ```

With Caching:
```
 hyperfine 'kubectl get svc'
Benchmark 1: kubectl get svc
  Time (mean ± σ):      5.329 s ±  0.017 s    [User: 0.056 s, System: 0.043 s]
  Range (min … max):    5.295 s …  5.353 s    10 runs
```

I also tested against other kubectl commands, such as `kubectl get pods` and saw nearly identical results.

If anyone's curious it can be setup with something like this:
```yaml
users:
- name: CLUSTER_ARN
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      command: /Users/Zach.Paden/.kube/eks-token-cache.sh
      args:
      - EKS_CLUSTER_NAME
      env:
      - name: AWS_PROFILE
        value: MY_AWS_PROFILE
```

TLDR: on average this improves the practical performance of kubectl commands by less than 14%

I'd love to here if there's something I missed, or, if anyone here has significantly faster kubectl commands when not using a cache.

**Comment by RyanFitzSimmonsAK** [comment]
> I didn't have this problem but recently, it takes almost 4 second to get the token. I don't know what has changed, was there any change recently ?
> 
> ```
> aws --version
> aws-cli/2.30.2 Python/3.13.7 Darwin/24.6.0 source/arm64
> ```

Hi @sov-msaeed, are you still experiencing these issues?  I'm attempting to reproduce these long operation times, and am consistently getting a response in ~0.2s, on both my local Windows machine and a Linux EC2 instance.  Do you have any additional environment details, and could you provide debug logs?  You can get debug logs by adding `--debug` to your command, and redacting any sensitive information.  Thanks!

```
2026-04-03 12:00:19,170 - MainThread - awscli.clidriver - DEBUG - CLI version: aws-cli/2.30.2 Python/3.13.7 Windows/11 exe/AMD64
2026-04-03 12:00:19,171 - MainThread - awscli.clidriver - DEBUG - Arguments entered to CLI: ['eks', 'get-token', '--debug']
...
2026-04-03 12:00:50,443 - MainThread - botocore.auth - DEBUG - CanonicalRequest:
```

**Comment by RyanFitzSimmonsAK** [comment]
Additionally, when you're getting these slow operation times, is `get-token` the first operation you're running after installation? Cold start time might be a possible reason if that's the case.
